cub/cub/thread/thread_reduce.cuh

```diff
@@ -43,199 +41,515 @@
   {
-    array[i] = input[i];
+#  pragma unroll
+    for (int j = 0; j + i < length; j += i * 3)
+    {
+      auto value = reduction_op(array[j], array[j + i]);
+      array[j]   = (j + i * 2 < length) ? reduction_op(value, array[j + i * 2]) : value;
+    }
   }
-  using DpxReduceOp   = cub_operator_to_dpx_t<ReductionOp, T>;
-  using SimdType      = ::cuda::std::pair<T, T>;
-  auto unsigned_input = reinterpret_cast<const unsigned*>(array);
-  auto simd_reduction = ThreadReduceSequential<length / 2>(unsigned_input, DpxReduceOp{});
-  auto simd_values    = ::cuda::std::bit_cast<SimdType>(simd_reduction);
-  auto ret_value      = reduction_op(simd_values.first, simd_values.second);
-  return (length % 2 == 0) ? ret_value : reduction_op(ret_value, input[length - 1]);
+  return array[0];
 }
 
-// DPX/Sequential dispatch
+/***********************************************************************************************************************
+ * SIMD Reduction
+ **********************************************************************************************************************/
+
+// never reached. Protect instantion of ThreadReduceSimd with arbitrary types and operators
 template <typename Input,
           typename ReductionOp,
-          typename ValueT = ::cuda::std::__remove_cvref_t<decltype(::cuda::std::declval<Input>()[0])>,
-          typename AccumT = ::cuda::std::__accumulator_t<ReductionOp, ValueT>,
-          _CUB_TEMPLATE_REQUIRES(enable_dpx_reduction<Input, ReductionOp, AccumT>())>
-_CCCL_NODISCARD _CCCL_DEVICE _CCCL_FORCEINLINE AccumT ThreadReduce(const Input& input, ReductionOp reduction_op)
+          _CUB_TEMPLATE_REQUIRES(!cub::internal::enable_generic_simd_reduction<Input, ReductionOp>())>
+_CCCL_NODISCARD _CCCL_DEVICE _CCCL_FORCEINLINE auto
+ThreadReduceSimd(const Input& input, ReductionOp) -> ::cuda::std::__remove_cvref_t<decltype(input[0])>
 {
-  static_assert(sizeof(Input) != sizeof(Input), "a");
-  static_assert(detail::has_subscript<Input>::value, "Input must support the subscript operator[]");
-  static_assert(detail::has_size<Input>::value, "Input must have the size() method");
-  static_assert(detail::has_binary_call_operator<ReductionOp, ValueT>::value,
-                "ReductionOp must have the binary call operator: operator(ValueT, ValueT)");
-  NV_IF_TARGET(NV_PROVIDES_SM_90,
-               (return ThreadReduceDpx(input, reduction_op);),
-               (return ThreadReduceSequential<AccumT>(input, reduction_op);))
+  assert(false);
+  return input[0];
 }
 
 template <typename Input,
           typename ReductionOp,
-          typename ValueT = ::cuda::std::__remove_cvref_t<decltype(::cuda::std::declval<Input>()[0])>,
-          typename AccumT = ::cuda::std::__accumulator_t<ReductionOp, ValueT>,
-          _CUB_TEMPLATE_REQUIRES(!enable_dpx_reduction<Input, ReductionOp, AccumT>())>
+          _CUB_TEMPLATE_REQUIRES(cub::internal::enable_generic_simd_reduction<Input, ReductionOp>())>
```

Comment:
I am not a fan of that macro, everybody looking into this knows SFINAE and this is not adding anything

Reply:
agree but this macro makes the code easier to read instead of a full SFINAE. I would like to keep it